### PR TITLE
Reduce allocations

### PR DESCRIPTION
**Pull request by RoosterDragon**
A couple more small tweaks to reduce allocations and thus reduce the work the GC has to do.

These changes save about ~8% in allocations during the main game loop in an AI heavy replay.

### CODE REVIEW

File: OpenRA.Mods.Common/OccupiedCellsExts.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common
+{
+	using OccupiedCells = IEnumerable<Pair<CPos, SubCell>>;
+
+	public static class OccupiedCellsExts
+	{
+		public static bool AnyIsExplored(this OccupiedCells cells, Player byPlayer)
```
Comment (pchote) [inline OpenRA.Mods.Common/OccupiedCellsExts.cs]: Why not `Shroud.IsExplored(OccupiedCells cells)`?

Comment (pchote) [inline OpenRA.Mods.Common/OccupiedCellsExts.cs]: (as an extension method in mod code)

**Comment by pchote** [comment]
Looks sensible overall, but I think it would be better to invert the extension methods, and also needs a rebase.

**Comment by RoosterDragon** [comment]
Rebased and inverted the extension methods.

**Comment by pchote** [comment]
:+1:

File: OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs [hunk 2]
```diff
@@ -64,8 +64,7 @@ bool IsVisibleInner(Actor self, Player byPlayer)
 		{
 			// If fog is disabled visibility is determined by shroud
 			if (!byPlayer.Shroud.FogEnabled)
-				return self.OccupiesSpace.OccupiedCells()
-					.Any(o => byPlayer.Shroud.IsExplored(o.First));
+				return byPlayer.Shroud.AnyIsExplored(self.OccupiesSpace.OccupiedCells());
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs]: Awkward, `AnyExplored` or `IsAnyExplored` would sound better.

File: OpenRA.Mods.Common/Traits/Modifiers/HiddenUnderFog.cs [hunk 3]
```diff
@@ -33,8 +33,7 @@ protected override bool IsVisibleInner(Actor self, Player byPlayer)
 				return base.IsVisibleInner(self, byPlayer);
 
 			if (Info.Type == VisibilityType.Footprint)
-				return self.OccupiesSpace.OccupiedCells()
-					.Any(o => byPlayer.Shroud.IsVisible(o.First));
+				return byPlayer.Shroud.AnyIsVisible(self.OccupiesSpace.OccupiedCells());
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Modifiers/HiddenUnderFog.cs]: See above.

File: OpenRA.Mods.Common/Traits/Modifiers/HiddenUnderShroud.cs [hunk 4]
```diff
@@ -42,8 +42,7 @@ public HiddenUnderShroud(HiddenUnderShroudInfo info)
 		protected virtual bool IsVisibleInner(Actor self, Player byPlayer)
 		{
 			if (Info.Type == VisibilityType.Footprint)
-				return self.OccupiesSpace.OccupiedCells()
-					.Any(o => byPlayer.Shroud.IsExplored(o.First));
+				return byPlayer.Shroud.AnyIsExplored(self.OccupiesSpace.OccupiedCells());
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Modifiers/HiddenUnderShroud.cs]: See above.

**Comment by atlimit8** [comment]
Otherwise, :+1:

**Comment by atlimit8** [comment]
`AnyVisible` and `AnyVisible` are preferred. They are sufficiently descriptive.
`IsAnyVisible` and `IsAnyVisible` are OK if `AnyVisible` and `AnyVisible` look like they are for enumerables.

**Comment by RoosterDragon** [comment]
Changed to `AnyExplored` and `AnyVisible`.

**Comment by atlimit8** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/cc998c71ca4e1357e5fafe3ed71874181f8e9f8f...58fbd154ca050c08abda097688100f2870add726)

**Pull request merged**